src/git.ts

```diff
@@ -0,0 +1,259 @@
+    }
+    return promisify(this._github.repos.get)({user: owner, repo: repo})
+        .then((response) => {
+          // TODO(usergenic): Patch to _handle_ redirects and/or include
+          // details in error messaging.  This was encountered because we
+          // tried to request Polymer/hydrolysis which has been renamed to
+          // Polymer/polymer-analyzer.
+          if (isRedirect(response)) {
+            console.log('Repo ${owner}/${repo} has moved permanently.');
+            console.log(response);
+            throw(`Repo ${owner}/${repo} could not be loaded.`);
+          }
+          return response;
+        });
+  }
+
+  /**
+   * @returns an array of repo (full_name) values for the given owner (which is either an
+   * org or user on github.)
+   */
+  async getRepoFullNames(owner: string): Promise<string[]> {
+    const names: string[] = [];
+
+    // Try to get the repo names assuming owner is an org.
+    const getFromOrg = promisify(this._github.repos.getFromOrg);
+    let pageSize = 50;
+    let page = 0;
+    let repos: GitHub.Repo[] = [];
+    const ownerRepoMap = new Map<string, GitHub.Repo>();
+    this._cache.repos.set(owner.toLowerCase(), ownerRepoMap);
+    do {
+      repos = await getFromOrg({org: owner, per_page: pageSize, page: page});
+      for (const repo of repos) {
+        names.push(repo.full_name);
+        ownerRepoMap.set(repo.name.toLowerCase(), repo);
+      }
+      ++page;
+    } while (repos.length > 0);
+
+    // TODO(usergenic): Update this function to support user repos as well as
+    // the org repos.
+    return names;
+  }
+
+  /**
+   * Given a nodegit repository, issue a git pull to bring it up to date.
+   */
+  async update(nodegitRepo: nodegit.Repository) {
+    await nodegitRepo.fetch('origin', this._cloneOptions.fetchOpts);
+  }
+
+  /**
+   * @returns the current user info from GitHub.  The current user is
+   * determined by the token being used.
+   */
+  async getCurrentUser(): Promise<GitHub.User> {
+    return promisify(this._github.user.get)({});
+  }
+}
+
+/**
+ * Checks out a branch with a given name on a repo.
+ *
+ * @returns the nodegit Branch object for the new branch.
+ */
+export async function checkout(
+    nodegitRepo: nodegit.Repository,
+    checkoutRef?: string): Promise<nodegit.Repository> {
+  const cwd = nodegitRepo.workdir();
+  const ref = typeof checkoutRef === 'string' ? checkoutRef : 'master';
+  return new Promise<nodegit.Repository>(
+      (resolve, reject) => (child_process.exec(
+          `git checkout ${ref}`, {cwd: cwd}, (error, stdout, stderr) => {
+            if (error) {
+              console.log(ref);
+              console.log(`Error checking out ${ref} in : ${cwd}`);
+            }
+            resolve(nodegitRepo);
+          })));
+}
+
+/**
+ * @returns a string representation of a RepoRef of the form:
+ *     "name:org/repo#ref"
+ */
+export function serializeGitHubRepoRef(repoRef: GitHubRepoRef): string {
```

Comment:
Will consider as we extend the behavior of the shorthand names, branch refs, tokens, repo specs, (or whatever we'll call them as a category after adding local paths support.)